In [1]:
import numpy as np
import pandas as pd

In [7]:
df = pd.DataFrame()
df['roll_one'] = np.random.randint(1, 7, 10_000)
df['roll_two'] = np.random.randint(1, 7, 10_000)
(df.roll_one == df.roll_two).mean()

0.1633

In [15]:
nrows = 10_000
ncols = 8

np.random.choice(['Heads', 'Tails'], size=80_000).reshape(nrows, ncols)

array([['Heads', 'Tails', 'Heads', ..., 'Heads', 'Heads', 'Heads'],
       ['Tails', 'Heads', 'Tails', ..., 'Heads', 'Heads', 'Heads'],
       ['Tails', 'Heads', 'Tails', ..., 'Heads', 'Tails', 'Heads'],
       ...,
       ['Heads', 'Heads', 'Tails', ..., 'Tails', 'Tails', 'Heads'],
       ['Tails', 'Heads', 'Tails', ..., 'Heads', 'Heads', 'Tails'],
       ['Heads', 'Tails', 'Heads', ..., 'Tails', 'Tails', 'Tails']],
      dtype='<U5')

In [13]:
flips = np.random.choice(['Heads', 'Tails'], size=(10_000, 8))
n_heads = (flips == 'Heads').sum(axis=1)
(n_heads == 3).mean()

0.2143

In [14]:
(n_heads > 3).mean()

0.6426

In [43]:
1_000 == 1000

True

In [40]:
simulations = np.random.choice(
    ['Web Dev', 'Data Science'],
    p=[.75, .25],
    size=(100_000, 2)
)
((simulations == 'Data Science').sum(axis=1) == 2).mean()

0.06293

In [35]:
((np.random.uniform(size=(100_000, 2)) <= .25).sum(axis=1) == 2).mean()

0.06255

In [51]:
purchases = np.random.normal(3, 1.5, size=(10_000, 4)).round()
purchases = np.where(purchases < 0, 0, purchases)
(purchases.sum(axis=1) <= 16).mean()

0.9354

In [53]:
np.array([1, 1.01, 1.4, 1.5, 1.9, 1.999999]).astype(int)

array([1, 1, 1, 1, 1, 1])

In [64]:
pd.options.display.max_rows = 50
# pd.DataFrame(np.random.rand(10_000, 10_000))

In [68]:
mens_heights = np.random.normal(178, 8, 10_000)
womens_heights = np.random.normal(170, 6, 10_000)

(womens_heights > mens_heights).mean()

0.2038

In [87]:
# True == corrupted download
n_students = 50
simulations = np.random.choice(
    [True, False],
    p=[1/250, 249/250],
    size=(10_000, n_students)
)
1 - simulations.any(axis=1).mean()

0.8213

In [88]:
# True == corrupted download
n_students = 100
simulations = np.random.choice(
    [True, False],
    p=[1/250, 249/250],
    size=(10_000, n_students)
)
1 - simulations.any(axis=1).mean()

0.6644

In [89]:
# True == corrupted download
n_students = 150
simulations = np.random.choice(
    [True, False],
    p=[1/250, 249/250],
    size=(10_000, n_students)
)
simulations.any(axis=1).mean()

0.4508

In [90]:
# True == corrupted download
n_students = 450
simulations = np.random.choice(
    [True, False],
    p=[1/250, 249/250],
    size=(10_000, n_students)
)
(simulations.sum(axis=1) == 0).mean()

0.1646

In [100]:
# True == corrupted download
n_students = 300
simulations = np.random.choice(
    [True, False],
    p=[1/250, 249/250],
    size=(10_000, n_students)
)
simulations.any(axis=1).mean()

0.7003

In [120]:
# If 450 students download anaconda, what is the likelihood that
# there are no errors in the first 150, but there's at least
# one error in the latter 300?
first_150_student_sims = np.random.choice(
    [True, False],
    p=[1/250, 249/250],
    size=(10_000, 150)
)
last_300_student_sims = np.random.choice(
    [True, False],
    p=[1/250, 249/250],
    size=(10_000, 300)
)

df = pd.DataFrame()
df['errors_in_first_150'] = first_150_student_sims.any(axis=1)
df['errors_in_last_300'] = last_300_student_sims.any(axis=1)
((~ df.errors_in_first_150) & (df.errors_in_last_300)).mean()

0.3803

In [99]:
~ pd.Series([True, False, True])

0    False
1     True
2    False
dtype: bool

P(at least one error in 150 students) = 0.4508
P(no errors in first 150) = 1 - .4508
P(at least one error in 300 students) = 0.7003

In [117]:
(1 - .4508) * .7003

0.38460476000000005

In [124]:
simulations = np.random.choice(
    ['Food Truck', 'No Food Truck'],
    p=[.7, .3],
    size=(10_000, 3)
)
(simulations == 'No Food Truck').all(axis=1).mean()

0.0268

In [167]:
simulations = np.random.choice(
    ['Food Truck', 'No Food Truck'],
    p=[.7, .3],
    size=(100_000, 5)
)
(simulations == 'Food Truck').any(axis=1).mean()

0.99777

In [168]:
10_000 / 52

192.30769230769232

In [177]:
simulations = np.random.randint(1, 366, size=(10_000, 23))
simulations = pd.DataFrame(simulations)
simulations
(simulations.apply(lambda birthdays: birthdays.nunique(), axis=1) < 23).mean()

0.5115

In [182]:
simulations = np.random.randint(1, 366, size=(10_000, 20))
simulations = pd.DataFrame(simulations)
simulations
(simulations.apply(lambda birthdays: birthdays.nunique(), axis=1) < 20).mean()

0.413

In [189]:
simulations = np.random.randint(1, 366, size=(10_000, 50))
simulations = pd.DataFrame(simulations)
simulations
(simulations.apply(lambda birthdays: birthdays.nunique(), axis=1) < 50).mean()

0.9715